File: sdk-python/src/custd/admin_offboarding.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, NotRequired, TypedDict, cast

from .client import AdminClient, AdminRequestOptions, quote_path


class OffboardingScheduleRequest(TypedDict):
    tenant_slug: str
    effective_at: str
    grace_period_days: int
    reason: str
    status: NotRequired[str]


class OffboardingSchedule(TypedDict, total=False):
    tenant_slug: str
    effective_at: str
    grace_period_days: int
    reason: str
    status: str
    updated_at: str


class OffboardingScheduleListResponse(TypedDict):
    schedules: list[OffboardingSchedule]


class OffboardingCancelRequest(TypedDict):
    reason: str


class OffboardingRequestCreate(TypedDict):
    confirmation: str


class OffboardingRequest(TypedDict, total=False):
    request_uuid: str
    state: str
    requested_at: str


class OffboardingPreviewStore(TypedDict, total=False):
    store: str
    kind: str
    retention_class: str
    estimated_count: int
    source_authority: str


class OffboardingPreviewResponse(TypedDict, total=False):
    request_uuid: str
    generated_at: str
    expires_at: str
    stores: list[OffboardingPreviewStore]
    exclusions: list[dict[str, object]]
    preview_inventory_digest: str
    complete: bool
    partial: bool


class OffboardingWaiver(TypedDict):
    role: str
    reason: str
    timestamp: NotRequired[str]


class OffboardingExecuteRequest(TypedDict):
    waiver: OffboardingWaiver


class OffboardingExportResponse(TypedDict):
    request_uuid: str
    checksum_sha256: str
    byte_size: int
    record_count: int
    generated_at: str
    expires_at: str
    preview_inventory_digest: str


class OffboardingDownloadResponse(TypedDict):
    request_uuid: str
    download_url: str
    checksum_sha256: str
    byte_size: int
    record_count: int
    generated_at: str
    expires_at: str
    preview_inventory_digest: str


class OffboardingReceiptPerStore(TypedDict):
    store: str
    retention_class: str
    deleted_count: int
    retained_exceptions_count: int


class OffboardingReceiptResponse(TypedDict, total=False):
    company_id: int
    requested_by_user_id: int | None
    requested_by_actor: str
    requested_at: str
    completed_at: str
    final_state: str
    per_store: list[OffboardingReceiptPerStore]
    waiver: OffboardingWaiver | None
    sha256: str


def _value(body: Mapping[str, Any], name: str, wire_name: str | None = None) -> Any:
    if name in body:
        return body[name]
    if wire_name is not None and wire_name in body:
        return body[wire_name]
    raise KeyError(name)


def _schedule_payload(body: OffboardingScheduleRequest) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "tenantSlug": _value(body, "tenant_slug", "tenantSlug"),
        "effectiveAt": _value(body, "effective_at", "effectiveAt"),
        "gracePeriodDays": _value(body, "grace_period_days", "gracePeriodDays"),
        "reason": _value(body, "reason"),
    }
    status = body.get("status")
    if status:
        payload["status"] = status
    return payload


def _request_from_wire(response: Mapping[str, Any]) -> OffboardingRequest:
    result: dict[str, Any] = {}
    _copy_value(result, "request_uuid", response, "requestUuid")
    _copy_value(result, "state", response, "state")
    _copy_value(result, "requested_at", response, "requestedAt")
    return cast(OffboardingRequest, result)


def _schedule_from_wire(response: Mapping[str, Any]) -> OffboardingSchedule:
    result: dict[str, Any] = {}
    _copy_value(result, "tenant_slug", response, "tenantSlug")
    _copy_value(result, "effective_at", response, "effectiveAt")
    _copy_value(result, "grace_period_days", response, "gracePeriodDays")
    _copy_value(result, "reason", response, "reason")
    _copy_value(result, "status", response, "status")
    _copy_value(result, "updated_at", response, "updatedAt")
    return cast(OffboardingSchedule, result)


def _preview_from_wire(response: Mapping[str, Any]) -> OffboardingPreviewResponse:
    result: dict[str, Any] = {}
    _copy_value(result, "request_uuid", response, "requestUuid")
    _copy_value(result, "generated_at", response, "generatedAt")
    _copy_value(result, "expires_at", response, "expiresAt")
    _copy_value(result, "preview_inventory_digest", response, "previewInventoryDigest")
    _copy_value(result, "complete", response, "complete")
    _copy_value(result, "partial", response, "partial")
    if "exclusions" in response:
        result["exclusions"] = response["exclusions"]
    stores = response.get("stores", [])
    result["stores"] = [_preview_store_from_wire(store) for store in _mapping_list(stores)]
    return cast(OffboardingPreviewResponse, result)


def _preview_store_from_wire(store: Mapping[str, Any]) -> OffboardingPreviewStore:
    result: dict[str, Any] = {}
    _copy_value(result, "store", store, "store")
    _copy_value(result, "kind", store, "kind")
    _copy_value(result, "retention_class", store, "retention_class")
    _copy_value(result, "estimated_count", store, "estimated_count")
    _copy_value(result, "source_authority", store, "source_authority")
    return cast(OffboardingPreviewStore, result)


def _export_from_wire(response: Mapping[str, Any]) -> OffboardingExportResponse:
    result: dict[str, Any] = {}
    _copy_value(result, "request_uuid", response, "requestUuid")
    _copy_value(result, "checksum_sha256", response, "checksumSha256")
    _copy_value(result, "byte_size", response, "byteSize")
    _copy_value(result, "record_count", response, "recordCount")
    _copy_value(result, "generated_at", response, "generatedAt")
    _copy_value(result, "expires_at", response, "expiresAt")
    _copy_value(result, "preview_inventory_digest", response, "previewInventoryDigest")
    return cast(OffboardingExportResponse, result)


def _download_from_wire(response: Mapping[str, Any]) -> OffboardingDownloadResponse:
    result: dict[str, Any] = {}
    _copy_value(result, "request_uuid", response, "requestUuid")
    _copy_value(result, "download_url", response, "downloadUrl")
    _copy_value(result, "checksum_sha256", response, "checksumSha256")
    _copy_value(result, "byte_size", response, "byteSize")
    _copy_value(result, "record_count", response, "recordCount")
    _copy_value(result, "generated_at", response, "generatedAt")
    _copy_value(result, "expires_at", response, "expiresAt")
    _copy_value(result, "preview_inventory_digest", response, "previewInventoryDigest")
    return cast(OffboardingDownloadResponse, result)


def _receipt_from_wire(response: Mapping[str, Any]) -> OffboardingReceiptResponse:
    result: dict[str, Any] = {}
    _copy_value(result, "company_id", response, "company_id")
    _copy_value(result, "requested_by_user_id", response, "requested_by_user_id")
    _copy_value(result, "requested_by_actor", response, "requested_by_actor")
    _copy_value(result, "requested_at", response, "requested_at")
    _copy_value(result, "completed_at", response, "completed_at")
    _copy_value(result, "final_state", response, "final_state")
    _copy_value(result, "sha256", response, "sha256")
    if "waiver" in response:
        result["waiver"] = response["waiver"]
    rows = response.get("per_store", [])
    result["per_store"] = [_receipt_store_from_wire(store) for store in _mapping_list(rows)]
    return cast(OffboardingReceiptResponse, result)


def _receipt_store_from_wire(store: Mapping[str, Any]) -> OffboardingReceiptPerStore:
    result: dict[str, Any] = {}
    _copy_value(result, "store", store, "store")
    _copy_value(result, "retention_class", store, "retention_class")
    _copy_value(result, "deleted_count", store, "deleted_count")
    _copy_value(result, "retained_exceptions_count", store, "retained_exceptions_count")
    return cast(OffboardingReceiptPerStore, result)


def _mapping_list(value: Any) -> list[Mapping[str, Any]]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, Mapping)]


def _copy_value(target: dict[str, Any], name: str, source: Mapping[str, Any], *wire_names: str) -> None:
    for wire_name in wire_names:
        if wire_name in source:
            target[name] = source[wire_name]
            return


class OffboardingClient:
    def __init__(self, admin: AdminClient) -> None:
        self._admin = admin

    def schedule(
        self,
        body: OffboardingScheduleRequest,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingSchedule:
        response = self._admin.request("POST", "/offboarding/schedules", _schedule_payload(body), options)
        return _schedule_from_wire(response)

    def list_schedules(self, options: AdminRequestOptions | None = None) -> OffboardingScheduleListResponse:
        response = self._admin.request("GET", "/offboarding/schedules", options=options)
        schedules = response.get("schedules", [])
        return {"schedules": [_schedule_from_wire(item) for item in _mapping_list(schedules)]}

    def get_schedule(
        self,
        tenant_slug: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingSchedule:
        response = self._admin.request("GET", f"/offboarding/schedules/{quote_path(tenant_slug)}", options=options)
        return _schedule_from_wire(response)

    def cancel_schedule(
        self,
        tenant_slug: str,
        body: OffboardingCancelRequest,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingSchedule:
        response = self._admin.request(
            "POST", f"/offboarding/schedules/{quote_path(tenant_slug)}/cancel", dict(body), options
        )
        return _schedule_from_wire(response)

    def request_offboarding(
        self,
        body: OffboardingRequestCreate,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        response = self._admin.request("POST", "/offboarding", dict(body), options)
        return _request_from_wire(response)

    def request(
        self,
        body: OffboardingRequestCreate,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        return self.request_offboarding(body, options)

    def get_request(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        response = self._admin.request("GET", f"/offboarding/{quote_path(request_uuid)}", options=options)
        return _request_from_wire(response)

    def cancel_request(
        self,
        request_uuid: str,
        body: OffboardingCancelRequest,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        response = self._admin.request("POST", f"/offboarding/{quote_path(request_uuid)}/cancel", dict(body), options)
        return _request_from_wire(response)

    def confirm_request(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        response = self._admin.request("POST", f"/offboarding/{quote_path(request_uuid)}/confirm", options=options)
        return _request_from_wire(response)

    def preview(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingPreviewResponse:
        response = self._admin.request(
            "POST", f"/offboarding/requests/{quote_path(request_uuid)}/preview", options=options
        )
        return _preview_from_wire(response)

    def export(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingExportResponse:
        response = self._admin.request(
            "POST", f"/offboarding/requests/{quote_path(request_uuid)}/export", options=options
        )
        return _export_from_wire(response)

    def download(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingDownloadResponse:
        response = self._admin.request(
            "GET", f"/offboarding/requests/{quote_path(request_uuid)}/download", options=options
        )
        return _download_from_wire(response)

    def acknowledge(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingRequest:
        response = self._admin.request(
            "POST", f"/offboarding/requests/{quote_path(request_uuid)}/acknowledge", options=options
        )
        return _request_from_wire(response)

    def execute(
        self,
        request_uuid: str,
        body: OffboardingExecuteRequest,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingReceiptResponse:
        waiver = body["waiver"]
        wire_body: dict[str, Any] = {
            "waiver_role": waiver["role"],
            "waiver_reason": waiver["reason"],
        }
        if waiver.get("timestamp"):
            wire_body["waiver_timestamp"] = waiver["timestamp"]
        response = self._admin.request(
            "POST", f"/offboarding/requests/{quote_path(request_uuid)}/execute", wire_body, options
        )
        return _receipt_from_wire(response)

    def retry(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingReceiptResponse:
        response = self._admin.request(
            "POST", f"/offboarding/requests/{quote_path(request_uuid)}/retry", options=options
        )
        return _receipt_from_wire(response)

    def receipt(
        self,
        request_uuid: str,
        options: AdminRequestOptions | None = None,
    ) -> OffboardingReceiptResponse:
        response = self._admin.request(
            "GET", f"/offboarding/requests/{quote_path(request_uuid)}/receipt", options=options
        )
        return _receipt_from_wire(response)

File: sdk-js/src/admin-offboarding.ts
// OffboardingClient owns the offboarding schedule and one-off request
// surfaces. Its types mirror the current admin-api JSON wire contract.

import type { RequestOptions } from "./index.js";

export type OffboardingSchedule = {
  tenantSlug: string;
  effectiveAt: string;
  gracePeriodDays: number;
  reason: string;
  status: string;
  updatedAt?: string;
};

export type OffboardingScheduleRequest = {
  tenantSlug: string;
  effectiveAt: string;
  gracePeriodDays: number;
  reason: string;
  status?: string;
};

export type OffboardingScheduleListResponse = {
  schedules: OffboardingSchedule[];
};

export type OffboardingCancelRequest = {
  reason: string;
};

// OffboardingRequest is the response shape for request, getRequest, and
// acknowledgement. The server calls the lifecycle field state (schedules
// use status).
export type OffboardingRequest = {
  requestUuid: string;
  state: string;
  requestedAt: string;
};

// OffboardingRequestCreate is the body for POST /offboarding. confirmation
// is the human-typed string the server compares against the tenant slug
// before accepting the destructive transition.
export type OffboardingRequestCreate = {
  confirmation: string;
};

// OffboardingPreviewStore is one row of the server-computed inventory. The
// SDK normalizes the server's snake_case durable fields for JavaScript callers.
export type OffboardingPreviewStore = {
  store: string;
  kind: string;
  retentionClass: string;
  estimatedCount: number;
  sourceAuthority?: string;
};

// OffboardingPreviewResponse is the body for POST
// /admin/offboarding/requests/{requestUuid}/preview.
export type OffboardingPreviewResponse = {
  requestUuid: string;
  generatedAt: string;
  expiresAt: string;
  stores: OffboardingPreviewStore[];
  exclusions?: Array<Record<string, unknown>>;
  previewInventoryDigest: string;
  complete: boolean;
  partial: boolean;
};

// OffboardingWaiver is the typed waiver the execute endpoint requires.
// role identifies the actor (e.g. client_owner); reason is the human-readable
// rationale. timestamp is server-stamped on accept.
export type OffboardingWaiver = {
  role: string;
  reason: string;
  timestamp?: string;
};

// OffboardingExecuteRequest is the body for POST
// /admin/offboarding/requests/{requestUuid}/execute. The server strictly
// decodes these top-level snake_case fields; waiver is not a nested object.
export type OffboardingExecuteRequest = {
  waiver: OffboardingWaiver;
};

// OffboardingExportResponse is the body for POST
// /admin/offboarding/requests/{requestUuid}/export.
export type OffboardingExportResponse = {
  requestUuid: string;
  checksumSha256: string;
  byteSize: number;
  recordCount: number;
  generatedAt: string;
  expiresAt: string;
  previewInventoryDigest: string;
};

// OffboardingDownloadResponse is the durable export descriptor returned by
// GET /admin/offboarding/requests/{requestUuid}/download. The downloadUrl is
// short-lived; callers must not log it or echo it into error messages. The
// remaining fields are the server's authoritative verification metadata.
export type OffboardingDownloadResponse = {
  requestUuid: string;
  downloadUrl: string;
  checksumSha256: string;
  byteSize: number;
  recordCount: number;
  generatedAt: string;
  expiresAt: string;
  previewInventoryDigest: string;
};

export type OffboardingAcknowledgeResponse = OffboardingRequest;

// Execute and retry return the same content-free receipt shape as GET receipt.
export type OffboardingExecuteResponse = OffboardingReceiptResponse;

export type OffboardingRetryResponse = OffboardingReceiptResponse;

// OffboardingReceiptPerStore is one row of the receipt's per-store summary.
// deletedCount is server-issued; retainedExceptionsCount covers legal holds
// and equivalent exclusions the SDK must not collapse.
export type OffboardingReceiptPerStore = {
  store: string;
  retentionClass: string;
  deletedCount: number;
  retainedExceptionsCount: number;
};

// OffboardingReceiptResponse is the body for GET
// /admin/offboarding/requests/{requestUuid}/receipt. finalState is the
// terminal state of the request; sha256 is the unkeyed integrity checksum the
// client must store alongside its offboarding record.
export type OffboardingReceiptResponse = {
  companyId: number;
  requestedByActor: string;
  requestedByUserId?: number | null;
  requestedAt: string;
  completedAt: string;
  finalState: string;
  perStore: OffboardingReceiptPerStore[];
  waiver?: OffboardingWaiver | null;
  sha256: string;
};

type AdminRequester = <T>(method: string, path: string, body?: unknown, options?: RequestOptions) => Promise<T>;

type WirePreviewStore = {
  store: string;
  kind: string;
  retention_class: string;
  estimated_count: number;
  source_authority?: string;
};

type WirePreviewResponse = Omit<OffboardingPreviewResponse, "stores"> & { stores: WirePreviewStore[] };
type WireReceiptPerStore = {
  store: string;
  retention_class: string;
  deleted_count: number;
  retained_exceptions_count: number;
};
type WireReceiptResponse = {
  company_id: number;
  requested_by_actor: string;
  requested_by_user_id?: number | null;
  requested_at: string;
  completed_at: string;
  final_state: string;
  per_store: WireReceiptPerStore[];
  waiver?: OffboardingWaiver | null;
  sha256: string;
};

function mapPreview(response: WirePreviewResponse): OffboardingPreviewResponse {
  return {
    ...response,
    stores: response.stores.map((store) => ({
      store: store.store,
      kind: store.kind,
      retentionClass: store.retention_class,
      estimatedCount: store.estimated_count,
      ...(store.source_authority ? { sourceAuthority: store.source_authority } : {}),
    })),
  };
}

function mapReceipt(response: WireReceiptResponse): OffboardingReceiptResponse {
  return {
    companyId: response.company_id,
    requestedByActor: response.requested_by_actor,
    requestedByUserId: response.requested_by_user_id,
    requestedAt: response.requested_at,
    completedAt: response.completed_at,
    finalState: response.final_state,
    perStore: response.per_store.map((store) => ({
      store: store.store,
      retentionClass: store.retention_class,
      deletedCount: store.deleted_count,
      retainedExceptionsCount: store.retained_exceptions_count,
    })),
    waiver: response.waiver,
    sha256: response.sha256,
  };
}

export class OffboardingClient {
  constructor(private readonly request: AdminRequester) {}

  // schedule writes a delayed offboarding schedule. tenantSlug is required and
  // must match the authenticated tenant scope.
  schedule(body: OffboardingScheduleRequest, options?: RequestOptions): Promise<OffboardingSchedule> {
    return this.request("POST", "/offboarding/schedules", body, options);
  }

  listSchedules(options?: RequestOptions): Promise<OffboardingScheduleListResponse> {
    return this.request("GET", "/offboarding/schedules", undefined, options);
  }

  // getSchedule reads the delayed offboarding schedule for a single tenant.
  // It targets the per-tenant route GET /offboarding/schedules/{tenantSlug},
  // which is distinct from the global listSchedules collection read.
  getSchedule(tenantSlug: string, options?: RequestOptions): Promise<OffboardingSchedule> {
    return this.request("GET", `/offboarding/schedules/${encodeURIComponent(tenantSlug)}`, undefined, options);
  }

  cancelSchedule(
    tenantSlug: string,
    body: OffboardingCancelRequest,
    options?: RequestOptions,
  ): Promise<OffboardingSchedule> {
    return this.request("POST", `/offboarding/schedules/${encodeURIComponent(tenantSlug)}/cancel`, body, options);
  }

  // request submits a one-off offboarding request for the effective tenant
  // via POST /offboarding. The confirmation field must match the tenant
  // slug the server reads from the auth context; mismatches fail with 400.
  requestOffboarding(body: OffboardingRequestCreate, options?: RequestOptions): Promise<OffboardingRequest> {
    return this.request("POST", "/offboarding", body, options);
  }

  getRequest(requestUuid: string, options?: RequestOptions): Promise<OffboardingRequest> {
    return this.request("GET", `/offboarding/${encodeURIComponent(requestUuid)}`, undefined, options);
  }

  cancelRequest(
    requestUuid: string,
    body: OffboardingCancelRequest,
    options?: RequestOptions,
  ): Promise<OffboardingRequest> {
    return this.request("POST", `/offboarding/${encodeURIComponent(requestUuid)}/cancel`, body, options);
  }

  confirmRequest(requestUuid: string, options?: RequestOptions): Promise<OffboardingRequest> {
    return this.request("POST", `/offboarding/${encodeURIComponent(requestUuid)}/confirm`, undefined, options);
  }

  // preview asks the server to compute the per-store inventory estimate for
  // the offboarding request. The result is server-issued and must be
  // surfaced verbatim; the SDK must not re-derive estimatedCount.
  async preview(requestUuid: string, options?: RequestOptions): Promise<OffboardingPreviewResponse> {
    const response = await this.request<WirePreviewResponse>(
      "POST",
      `/offboarding/requests/${encodeURIComponent(requestUuid)}/preview`,
      undefined,
      options,
    );
    return mapPreview(response);
  }

  // export triggers the destructive export packaging for a request. The
  // response is the per-request artifact metadata; the download URL is
  // fetched separately via download.
  export(requestUuid: string, options?: RequestOptions): Promise<OffboardingExportResponse> {
    return this.request("POST", `/offboarding/requests/${encodeURIComponent(requestUuid)}/export`, undefined, options);
  }

  // download returns the durable export descriptor and a short-lived signed
  // URL for the offboarding artifact. The downloadUrl is sensitive; callers
  // must not log it or echo it into error messages.
  download(requestUuid: string, options?: RequestOptions): Promise<OffboardingDownloadResponse> {
    return this.request("GET", `/offboarding/requests/${encodeURIComponent(requestUuid)}/download`, undefined, options);
  }

  // acknowledge records that the export was downloaded successfully and its
  // inventory was confirmed. Never call it merely because preview succeeded.
  acknowledge(requestUuid: string, options?: RequestOptions): Promise<OffboardingAcknowledgeResponse> {
    return this.request(
      "POST",
      `/offboarding/requests/${encodeURIComponent(requestUuid)}/acknowledge`,
      undefined,
      options,
    );
  }

  // execute triggers the destructive phase. The server requires a non-empty
  // waiver.role; an empty waiver returns 400 waiver_required, which the
  // SDK surfaces without retry.
  execute(
    requestUuid: string,
    body: OffboardingExecuteRequest,
    options?: RequestOptions,
  ): Promise<OffboardingExecuteResponse> {
    const wireBody = {
      waiver_role: body.waiver.role,
      waiver_reason: body.waiver.reason,
      ...(body.waiver.timestamp ? { waiver_timestamp: body.waiver.timestamp } : {}),
    };
    return this.request<WireReceiptResponse>(
      "POST",
      `/offboarding/requests/${encodeURIComponent(requestUuid)}/execute`,
      wireBody,
      options,
    ).then(mapReceipt);
  }

  // retry re-arms an offboarding request that previously failed. The server
  // decides whether the request is retryable; the SDK does not pre-filter.
  retry(requestUuid: string, options?: RequestOptions): Promise<OffboardingRetryResponse> {
    return this.request<WireReceiptResponse>(
      "POST",
      `/offboarding/requests/${encodeURIComponent(requestUuid)}/retry`,
      undefined,
      options,
    ).then(mapReceipt);
  }

  // receipt returns the terminal offboarding receipt for a request. The
  // sha256 is an unkeyed integrity checksum the client must retain alongside
  // its offboarding record; it is not an authenticity signature.
  receipt(requestUuid: string, options?: RequestOptions): Promise<OffboardingReceiptResponse> {
    return this.request<WireReceiptResponse>(
      "GET",
      `/offboarding/requests/${encodeURIComponent(requestUuid)}/receipt`,
      undefined,
      options,
    ).then(mapReceipt);
  }
}
